docs/tutorial/reduction.rst

```diff
@@ -0,0 +1,720 @@
+                       gid = bid * blockDim.x + tid;
+
+        // Read input from front buffer to shared
+        shared[tid] = read_global_safe(gid);
+        __syncthreads();
+
+        // Shared reduction
+        for (uint32_t i = 1; i < blockDim.x; i *= 2)
+        {
+            if (tid % (2 * i) == 0)
+                shared[tid] = op(shared[tid], shared[tid + i]);
+            __syncthreads();
+        }
+
+        // Write result from shared to back buffer
+        if (tid == 0)
+            back[bid] = shared[0];
+    }
+
+While the ``tid % (2 * i) == 0`` indexing scheme yields correct results, it also leads to high thread divergence. Thread divergence indicates the event when the threads in a warp diverge, which implies that the threads have to execute different instructions in a given clock cycle. This is easily manifested using ``if-else`` statements as shown here, but can also be manifested as ``for`` loop dependent on thread id lengths. Even though the number of active threads participating in the reduction reduces, warps remain active for longer than necessary, as at least one lane in a warp hits the ``if`` statement.
+
+Reducing thread divergence
+--------------------------
+
+You can reduce divergence by keeping dataflow between memory addresses identical but reassigning the thread ids.
+
+.. figure:: ../data/tutorial/reduction/reduced_divergence_reduction.svg
+  :alt: Diagram demonstrating reduced divergence reduction
+
+.. note::
+
+    For those less proficient in reading Git diffs, the following code segments show changes between the two versions of a file. Lines highlighted in red are being removed or changed while lines highlighted in green are being introduced.
+
+.. code-block:: diff
+
+    // Shared reduction
+    for (uint32_t i = 1; i < blockDim.x; i *= 2)
+    {
+    -    if (tid % (2 * i) == 0)
+    -        shared[tid] = op(shared[tid], shared[tid + i]);
+    +    if (uint32_t j = 2 * i * tid; j < blockDim.x)
+    +        shared[j] = op(shared[j], shared[j + i]);
+        __syncthreads();
+    }
+
+This way inactive threads start accumulating uniformly towards the higher thread id index range and might uniformly skip to ``__syncthreads()``. However, this introduces a bank conflicts issue.
+
+Resolving bank conflicts
+------------------------
+
+Both AMD and NVIDIA implement shared memory in the hardware by organizing storage into banks of various sizes. On the AMD hardware, this hardware element is known as Local Data Share (LDS). On NVIDIA hardware, it's implemented using the same silicon as the L1 data cache. You can think of shared memory  as a striped 2-dimensional range of memory. Shared memory bank's count, width, and depth depend on the architecture. A bank conflict occurs when different threads in a warp access the same bank during the same operation. In this case, the hardware prevents the attempted concurrent accesses to the same bank by converting them into serial accesses.
```

Comment:
```suggestion
Both AMD and NVIDIA implement shared memory in the hardware by organizing storage into banks of various sizes. This hardware element is known as Local Data Share (LDS) on AMD hardware. On NVIDIA hardware, it's implemented using the same silicon as the L1 data cache. You can think of shared memory as a striped 2-dimensional range of memory. Shared memory bank's count, width, and depth depend on the architecture. A bank conflict occurs when different threads in a warp access the same bank during the same operation. In this case, the hardware prevents the attempted concurrent accesses to the same bank by converting them into serial accesses.
```

Reply:
Something wrong with this sentence:
Shared memory bank's count, width, and depth depend on the architecture.